Add Slime Ball max code

bsavage81 opened the pull request:
Adds code to increase the max slime balls per day, along with a GMCM menu.

bsavage81 committed:
Add Slime Ball max code

Adds code to increase the max slime balls per day, along with a GMCM menu.

Pathoschild commented on ChangeSlimeHutchLimit/ModEntry.cs:
```diff
@@ -11,19 +16,134 @@ public class ModEntry : Mod
+            }
+        }
+
+        return true;
```
Won't this result in the max limit being additive to the vanilla limit? For example, let's say you have 20 slimes, which would normally produce 4 slime balls. If you set the max slime balls to 6, that would spawn up to 10 slime balls per day (6 from the patch and 4 from the vanilla code).

bsavage81 replied:
As far as I can tell it doesn't do this. I have been playing with it for a few weeks now and I am pretty sure it is not additive

Pathoschild replied:
In a prefix patch, [`return true` means the original method should run too](https://harmony.pardeike.net/articles/patching-prefix.html#changing-the-result-and-skipping-the-original). So this patch spawns the desired number of slimes, then it runs the vanilla method which spawns slimes again.

bsavage81 replied:
Ok, so should we switch it to a postfix?

bsavage81 replied:
or am I overthinking it, and I should just return false

bsavage81 replied:
Went ahead and updated the code to be additive to the vanilla method

Pathoschild commented on ChangeSlimeHutchLimit/ModEntry.cs:
```diff
@@ -11,19 +16,134 @@ public class ModEntry : Mod
+
+        // Respect the user-defined maximum daily limit, if set
+        int maxBallsToPlace = _config.MaxDailySlimeBalls > 0
+            ? Math.Min(slimeBasedLimit, _config.MaxDailySlimeBalls)
```
It's probably confusing that the max slime balls is still limited to the number of water troughs, since most players won't have a custom slime hutch map. Even if they set the max slime balls to 100, it'll still be limited to 4 since there's only 4 water troughs.

bsavage81 replied:
This is following the vanilla logic that the troughs need to be watered in order for slime balls to appear. If you have 100 slime, and only 1 trough watered, it will register 25 slimes and divide that by 5 to get you 5 slime balls. If all 4 troughs are watered, you will register all 100/5 resulting in 25 slime balls. Basically the code maintains the vanilla reasoning of 25% for each watered trough.

Pathoschild commented on ChangeSlimeHutchLimit/ModEntry.cs:
```diff
@@ -11,19 +16,134 @@ public class ModEntry : Mod
+                Object slimeBall = ItemRegistry.Create<Object>("(BC)56"); // Slime Ball ID
+#pragma warning disable AvoidNetField // Avoid Netcode types when possible
+                slimeBall.fragility.Value = 2;
+#pragma warning restore AvoidNetField // Avoid Netcode types when possible
```
We can set `slimeBall.Fragility = 2` instead to avoid the net field warning.

bsavage81 replied:
Implemented

Pathoschild commented on ChangeSlimeHutchLimit/ModEntry.cs:
```diff
@@ -11,19 +16,134 @@ public class ModEntry : Mod
+        this.SetupGMCM();
+    }
+
+    private void SetupGMCM()
```
Could you split the Generic Mod Config Menu integration into a separate PR? That way it can be reviewed/tweaked/merged separately while we discuss the slime ball changes.

Pathoschild replied:
Done via PR #100.

Pathoschild reviewed (changes requested):
Thanks for the pull request! I added a few comments inline to discuss.

bsavage81 committed:
Updated changes in discussion

bsavage81 reviewed (commented):
I don't use Github much, so not sure how to split a PR into two. Let me know if you see any other issues.

bsavage81 committed:
Merge branch 'ChroniclerCherry:develop' into develop

bsavage81 committed:
Update ModEntry.cs

bsavage81 committed:
Updated code for slime ball limit

Updated the code for the slime limit so that it is additive to the vanilla behavior

bsavage81 committed:
Update MaxBallsToPlace

Added a subtraction for the waterSpots that will generate with the vanilla code so that the Max set in the config will be accurate.

bsavage81 commented:
Closed in favor of changes in [PR 106](https://github.com/ChroniclerCherry/stardew-valley-mods/pull/106)
